src/main/java/gregtech/common/entities/EntityGTExplosive.java

```diff
@@ -0,0 +1,136 @@
+        super(world, x, y, z, exploder);
+    }
+
+    @SuppressWarnings("unused")
+    public EntityGTExplosive(World world) {
+        super(world);
+    }
+
+    /**
+     * @return The strength of the explosive.
+     */
+    protected abstract float getStrength();
+
+    /**
+     * @return Whether to drop all blocks, or use default logic
+     */
+    public abstract boolean dropsAllBlocks();
+
+    /**
+     * @return The range of the explosive, if {@link #dropsAllBlocks} is true.
+     */
+    protected int getRange() {
+        return 2;
+    }
+
+    /**
+     * @return The block state of the block this explosion entity is created by.
+     */
+    public abstract @NotNull IBlockState getExplosiveState();
+
+    @Override
+    public void onUpdate() {
+        this.prevPosX = this.posX;
+        this.prevPosY = this.posY;
+        this.prevPosZ = this.posZ;
+        if (!this.hasNoGravity()) {
+            this.motionY -= 0.03999999910593033D;
+        }
+
+        this.move(MoverType.SELF, this.motionX, this.motionY, this.motionZ);
+        this.motionX *= 0.9800000190734863D;
+        this.motionY *= 0.9800000190734863D;
+        this.motionZ *= 0.9800000190734863D;
+        if (this.onGround) {
+            this.motionX *= 0.699999988079071D;
+            this.motionZ *= 0.699999988079071D;
+            this.motionY *= -0.5D;
+        }
+
+        setFuse(this.getFuse() - 1);
+        if (this.getFuse() <= 0) {
+            this.setDead();
+            if (!this.world.isRemote) {
+                this.explodeTNT();
+            }
+        } else {
+            this.handleWaterMovement();
+            this.world.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, this.posX, this.posY + 0.5D, this.posZ, 0.0D, 0.0D,
+                    0.0D);
+        }
+    }
+
+    protected void explodeTNT() {
+        this.world.createExplosion(this, this.posX, this.posY + (double) (this.height / 16.0F), this.posZ,
+                getStrength(), !dropsAllBlocks());
+
+        // If we don't drop all blocks, then skip the drop capture logic
+        if (!dropsAllBlocks()) return;
+
+        // Create the fake explosion but don't destroy any blocks in water, per MC behavior
+        if (this.inWater) return;
+
+        EntityPlayer player = GregFakePlayer.get((WorldServer) world);
+
+        int range = getRange();
+        for (BlockPos pos : BlockPos.getAllInBox(this.getPosition().add(-range, -range, -range),
+                this.getPosition().add(range, range, range))) {
+            IBlockState state = world.getBlockState(pos);
+
+            if (state.getMaterial() == Material.AIR) continue;
+            if (state.getMaterial() == Material.WATER || state.getMaterial() == Material.LAVA) continue;
+
+            float hardness = state.getBlockHardness(world, pos);
+            float resistance = state.getBlock().getExplosionResistance(player);
+
+            if (hardness >= 0.0f && resistance < 100 && world.isBlockModifiable(player, pos)) {
```

Comment:
should this check for if the resistance is more than the strength (or some factor of the strength)? I am not really sure how regular tnt works

Reply:
This works differently than normal TNT in that it guarantees drops so it doesn't use the same checks. But this value was chosen to break most blocks but not things like liquid blocks (resistance 100) and obsidian (1200) and similar high explosion resistance blocks